src/cls/IPM/General/History.cls

```diff
@@ -0,0 +1,394 @@
+    Set diff = diff # 3600
+    Set diffMinutes = diff \ 60
+    Set diffSeconds = diff # 60
+
+    Set duration = $Select(diffDays: diffDays _ "D ", 1:"") 
+        _$Select(diffHours: diffHours _ "h ", 1:"")
+        _$Select(diffMinutes: diffMinutes _ "m ", 1:"")
+        _$Select(diffSeconds: diffSeconds _ "s ", 1:"")
+
+    If duration = "" {
+        Set duration = "0s"
+    }
+
+    Quit rs.%Get("TimeStart") _ " (" _ $Zstrip(duration, "<>W") _ ")"
+}
+
+ClassMethod SourceToString(rs As %SQL.StatementResult, detail As %Boolean = 0) As %String [ Internal, Private ]
+{
+    If rs.%Get("Action") '= "install" {
+        Quit "n/a"
+    }
+    If rs.%Get("SourceName") = "" {
+        Quit "<no source set>"
+    }
+    If detail {
+        Quit rs.%Get("SourceName") _ " (" _ rs.%Get("SourceMoniker") _ "): " _ rs.%Get("SourceDetails")
+    } Else {
+        Quit rs.%Get("SourceName")
+    }
+}
+
+ClassMethod SuccessToString(rs As %SQL.StatementResult, detail As %Boolean = 0) As %String [ Internal, Private ]
+{
+    Set success = rs.%Get("Success")
+    If detail {
+        Quit success
+    }
+    Quit $Select($$$ISOK(success): "Success", 1:$System.Status.GetErrorText(success))
+}
+
+ClassMethod CommittedToString(rs As %SQL.StatementResult, detail As %Boolean = 0) As %String [ Internal, Private ]
+{
+    Quit $Select(rs.%Get("Committed"): "Committed", 1: "Uncommitted")
+}
+
+ClassMethod DisplayQueryResult(rs As %SQL.StatementResult)
+{
+    // TODO improve the display output
+    If (rs = $$$NULLOREF) {
+        Write !, "IPM is not enabled for this namespace"
+        Quit
+    }
+
+    set found = 0
+    While rs.%Next() {
+        If 'found {
+            Set found = 1
+            Do ..ShowColumns(rs, "header")
+            Write !
+        }
+        Do ..ShowColumns(rs, 0)
+    }
+    Write !
+}
+
+ClassMethod DisplayArray(array As %Library.ArrayOfObjects)
+{
+    Set ns = ""
+    For {
+        Set rs = array.GetNext(.ns)
+        If rs = "" {
+            Quit
+        }
+        Write !, "Namespace: ", ns
+        Do ..DisplayQueryResult(rs)
+    }
+}
+
+Storage Default
+{
+<Data name="HistoryDefaultData">
+<Value name="1">
+<Value>%%CLASSNAME</Value>
+</Value>
+<Value name="2">
+<Value>Action</Value>
+</Value>
+<Value name="3">
+<Value>Package</Value>
+</Value>
+<Value name="4">
+<Value>Version</Value>
+</Value>
+<Value name="5">
+<Value>SourceName</Value>
+</Value>
+<Value name="6">
+<Value>SourceMoniker</Value>
+</Value>
+<Value name="7">
+<Value>SourceDetails</Value>
+</Value>
+<Value name="8">
+<Value>TimeStart</Value>
+</Value>
+<Value name="9">
+<Value>TimeEnd</Value>
+</Value>
+<Value name="10">
+<Value>NameSpace</Value>
+</Value>
+<Value name="11">
+<Value>UserName</Value>
+</Value>
+<Value name="12">
+<Value>Success</Value>
+</Value>
+<Value name="13">
+<Value>Committed</Value>
+</Value>
+<Value name="14">
+<Value>CommandString</Value>
+</Value>
+</Data>
+<DataLocation>^%IPM.General.HistoryD</DataLocation>
```

Comment:
How is this mapped?

Reply:
The data is mapped to each namespace where %IPM package is mapped. Do we want to separate them?